compress/compress.go

```diff
@@ -0,0 +1,477 @@
+		bit := d.ReadBit()
+		xorLength = xorLength | (bit << i)
+		i--
+	}
+
+	j := uint64(xorLength)
+	var xor uint64
+	xor = 0
+	for uint32(j) != MaxUint32 {
+		bit := d.ReadBit()
+		xor = xor | (uint64(bit) << j)
+		j--
+	}
+
+	var rebuiltNumber uint64
+	rebuiltNumber = 0
+	previousBits := math.Float64bits(previous)
+	xor = xor << (64 - leadingZeros - xorLength)
+	rebuiltNumber = previousBits ^ xor
+
+	var buffer *bytes.Buffer
+	buffer = new(bytes.Buffer)
+	binary.Write(buffer, binary.BigEndian, rebuiltNumber)
+
+	b := buffer.Bytes()
+	readbuf := bytes.NewReader(b)
+	var result float64
+	err := binary.Read(readbuf, binary.BigEndian, &result)
+	if err != nil {
+		panic("WUT")
+	}
+
+	d.lengthOfWindow = uint64(xorLength)
+	d.leadingZeroWindowLength = uint64(leadingZeros)
+
+	return result
+}
+
+func (c *CompressionBuffer) Bytes() []byte {
+	if !c.finalized {
+		panic("Attempted to read bytes from an unfinalized compression buffer")
+	}
+	return c.buffer.Bytes()
+}
+
+func NewCompressionBuffer() CompressionBuffer {
+	return CompressionBuffer{
+		buffer:    new(bytes.Buffer),
+		position:  0,
+		current:   0,
+		firstPass: true,
+	}
+}
+
+func (c *CompressionBuffer) writeFirst(x float64) {
+	err := binary.Write(c.buffer, binary.BigEndian, x)
+	if err != nil {
+		panic("WUT")
+	}
+}
+
+func (c *CompressionBuffer) writeOne() {
+	c.current = c.current << uint64(1)
+	c.current = c.current | uint64(1)
+	c.position++
+	c.fixup()
+}
+
+func (c *CompressionBuffer) writeZero() {
+	c.current = c.current << 1
+	c.position++
+	c.fixup()
+}
+
+func (c *CompressionBuffer) fixup() {
+	if c.position == uint32(64) {
+		err := binary.Write(c.buffer, binary.BigEndian, c.current)
+		if err != nil {
+			panic("WUT")
+		}
+		c.position = 0
+		c.current = uint64(0)
+	}
+}
+
+func (c *CompressionBuffer) encodeMeaningfulXOR(x uint64) {
+	leadingZeros := leadingZeros64(x)
+	trailingZeros := trailingZeros64(x)
+	length := uint64(64) - uint64(leadingZeros) - uint64(trailingZeros)
+
+	c.writeOne()
+
+	//Case A: The meaningful bits of this block have the same # of leading
+	//zeros and field size as the previous compressed float.
+	if uint64(leadingZeros) == c.leadingZeroWindowLength && uint64(length) == c.lengthOfWindow && c.lengthOfWindow != 0 {
+		c.writeZero()
+		meaningfulRegion := x >> (64 - c.leadingZeroWindowLength - c.lengthOfWindow)
+		c.writeMeaningfulPartialXOR(c.lengthOfWindow, meaningfulRegion)
+	} else {
+		//Case B: This float has a different # of leading zeros and/or size
+		//than the previous one.
+		c.writeOne()
+		meaningfulRegion := (x >> trailingZeros)
+		c.writeMeaningfulXOR(uint64(leadingZeros), length, meaningfulRegion)
+
+		c.lengthOfWindow = length
+		c.leadingZeroWindowLength = uint64(leadingZeros)
+	}
+}
+
+func (c *CompressionBuffer) writeMeaningfulPartialXOR(length uint64, meaningful uint64) {
+	i := uint32(length)
+	for i != MaxUint32 {
+		bit := (meaningful >> i) & 1
+		if bit == 1 {
+			c.writeOne()
+		} else {
+			c.writeZero()
+		}
```

Comment:
This same sort of code appears several times. I think it would make sense to make a function:

```
func (c *CompressionBuffer) writeBits(value uint64, n uint32) {
    i := n
    for i != MaxUint32 {
        bit := (value >> i) & 1
        if bit == 1 {
            c.writeOne()
        } else {
            c.writeZero()
        }
        i--
    }
}
```

which will hopefully make it easier to understand (by abstracting away some of these lower-level details).

Actually, it might also be helpful to have

```
func (c *CompressionBuffer) writeBit(bit bool) {
    if bit {
        c.writeOne()
    } else {
        c.writeZero()
    }
}
```